### PR TITLE
Connect: Show config file errors

### PR DESCRIPTION
**Pull request by gzdunek**
Part of https://github.com/gravitational/teleport/issues/21914

I changed `getStoredConfigErrors` (now `getConfigError`) to return a union of two types of errors: `validation` (Zod errors) and `file-loading` (IO and JSON syntax errors).  

I also noticed that when the config file is corrupted, updating its contents (for example by linking to the JSON schema) removes all values from it. That is because when we can’t read the file, our logic creates an empty object in memory, which is then written to disk with the first update. To remedy this, I created a new flag: `discardUpdatesWhenLoadingFileFailed` (a better name is welcome) for the `FileStorage`. Enabling it discards all updates to the file. We don't want to enable it for all the files; for app state or json schema it is okay to overwrite them.

Technically, thanks to this flag, we don't have to hide the usage reporting dialog (normally a response from the user would overwrite the file), but I decided to do it because:
- we won't store the response anyway 
- the error notification is covered by the dialog backdrop

If you think that hiding the dialog is not needed then I am happy to discuss it.

Other than that, I refactored `FileStorage` to use async FS functions (suggested by @ravicious). 

Example notification:
image

### CODE REVIEW

File: web/packages/teleterm/src/services/fileStorage/fileStorage.ts [hunk 1]
```diff
@@ -37,25 +37,37 @@ export interface FileStorage {
...
+  getFileLoadingError(): Error | undefined;
 }
 
-export function createFileStorage(opts: {
+export async function createFileStorage(opts: {
   filePath: string;
   debounceWrites: boolean;
-}): FileStorage {
+  /** Prevents state updates when the file has not been loaded correctly, so its content will not be overwritten. */
+  discardUpdatesWhenLoadingFileFailed?: boolean;
```
Comment (ravicious) [inline web/packages/teleterm/src/services/fileStorage/fileStorage.ts:49]: I feel like this flag is where our past design decisions came back to bite us. 😶‍🌫️

Instead of adding a new flag, I feel like it'd be better to pass an alternative implementation of file storage that just ignores all writes and returns undefined for all reads. If you squint really hard, adding flags on top of each other is similar to using class inheritance to implement different behavior (remember that Sandi Metz video I linked to?).

`debounceWrites` as a flag just doesn't make any sense when `discardUpdatesWhenLoadingFileFailed` is set to true because then `debounceWrites` has no effect. Yet it's totally fine to pass `debounceWrites: true, discardUpdatesWhenLoadingFileFailed: true`. Instead of providing complex types to work our way around this, we could use a different implementation if there was an error.

Alas, we cannot really do that here as the error thrown during the file load is an integral part of a file storage and not a separate value that can be passed independently of the file storage that's being used.

Similar to my previous comment, I don't feel like it makes sense to refactor all of this now so I'm just leaving this comment for future consideration. Before we add a third flag or have to implement a similar object in the future, let's consider other solutions.

Comment (ravicious) [inline web/packages/teleterm/src/services/fileStorage/fileStorage.ts:49]: As for the name, maybe `discardUpdatesOnLoadError`?

File: web/packages/teleterm/src/ui/initUi.ts [hunk 2]
```diff
@@ -43,34 +43,47 @@ export async function initUi(ctx: IAppContext): Promise<void> {
...
-        href: 'https://goteleport.com/docs/connect-your-client/teleport-connect/',
-        text: 'See the config file documentation',
-      },
-    });
+  const configError = configService.getConfigError();
+  if (configError) {
+    switch (configError.source) {
+      case 'file-loading': {
+        notificationsService.notifyError({
+          title: 'Failed to load config file, using the default config',
```
Comment (ravicious) [inline web/packages/teleterm/src/ui/initUi.ts:62]: Either use `the` in both cases or not at all. 😏

How about this:

```typescript
          title: 'Failed to load the config file',
          description: `Using the default config instead.\n${configError.error}`,
```

File: web/packages/teleterm/src/ui/services/usage/setUpUsageReporting.ts [hunk 3]
```diff
@@ -29,6 +29,11 @@ export async function setUpUsageReporting(
     return;
   }
 
+  if (configService.getConfigError()?.source === 'file-loading') {
+    // do not show the dialog, response cannot be saved to the file
+    return;
```
Comment (ravicious) [inline web/packages/teleterm/src/ui/services/usage/setUpUsageReporting.ts:34]: This makes sense to me, the main argument being what you mentioned – if the response cannot be saved then it doesn't make sense to ask the question anyway.

File: web/packages/shared/components/Notification/Notification.tsx [hunk 4]
```diff
@@ -150,7 +150,7 @@ function getRenderedContent(
           `}
         >
           <Text
-            fontSize={14}
+            fontSize={13}
```
Comment (ravicious) [inline web/packages/shared/components/Notification/Notification.tsx:153]: This is a shared component now so we shouldn't really make sneaky changes like that, I'd at least put it in a separate PR. This way the rest of the team has a chance to review it and it can also be backported independent of config changes (rather than having different notification font sizes between versions). I do agree that it looks better though.

File: web/packages/teleterm/src/main.ts [hunk 5]
```diff
@@ -39,23 +39,15 @@ if (app.requestSingleInstanceLock()) {
   app.exit(1);
 }
 
-function initializeApp(): void {
+async function initializeApp(): Promise<void> {
```
Comment (ravicious) [inline web/packages/teleterm/src/main.ts:42]: Thanks for the refactor. Did you run into problems with the `writeSync` function? Is it used anywhere else other than when quitting the app?

File: web/packages/teleterm/src/services/fileStorage/fileStorage.ts [hunk 6]
```diff
@@ -91,19 +110,38 @@ export function createFileStorage(opts: {
...
+): Promise<{ state: any; error: Error | undefined }> {
   try {
-    if (!existsSync(filePath)) {
-      writeFileSync(filePath, '{}');
-    }
-
-    return JSON.parse(readFileSync(filePath, { encoding: 'utf-8' }));
+    const file = await readOrCreateFile(filePath);
+    return {
+      state: JSON.parse(file),
```
Comment (ravicious) [inline web/packages/teleterm/src/services/fileStorage/fileStorage.ts:123]: Also, it's a shame that Node.js doesn't have a dedicated function for parsing JSON files because the errors `JSON.parse` returns are rather unhelpful. It mentions the position in the string rather than the file which would be much more helpful.

File: web/packages/teleterm/src/services/fileStorage/fileStorage.ts [hunk 7]
```diff
@@ -91,19 +110,38 @@ export function createFileStorage(opts: {
...
+      error: undefined,
+    };
   } catch (error) {
     logger.error(`Cannot read ${filePath} file`, error);
-    return {};
+    return {
+      state: {},
+      error: error,
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply web/packages/teleterm/src/services/fileStorage/fileStorage.ts]: Sounds good, done.

File: web/packages/teleterm/src/services/config/configService.ts [hunk 8]
```diff
@@ -31,14 +31,31 @@ import {
...
 
   set<K extends keyof AppConfig>(key: K, value: AppConfig[K]): void;
 
-  getStoredConfigErrors(): ZodIssue[] | undefined;
+  /**
+   * Returns validation errors or an error that occurred during loading the config file (this means IO and syntax errors).
+   * If validation errors occur, the incorrect values are replaced with the defaults.
+   * In case of an error coming from loading the file, all values are replaced with the defaults.
+   * */
+  getConfigError(): ConfigError | undefined;
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply web/packages/teleterm/src/services/config/configService.ts:62]: Done

File: web/packages/teleterm/src/services/fileStorage/fileStorage.ts [hunk 9]
```diff
@@ -64,19 +76,26 @@ export function createFileStorage(opts: {
...
+    if (discardUpdates) {
+      return;
+    }
     state = json;
     stringifyAndWrite();
   }
 
+  function get<T>(key?: string): T {
+    return key ? state[key] : (state as T);
+  }
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply web/packages/teleterm/src/services/fileStorage/fileStorage.ts]: Absolutely agree, `get` in its current form doesn't make any sense, it only makes casting a little bit easier. 
I applied your patch, it is still better than what we had. 

>well, it'd be just as unsound as anything else as we don't verify that the contents of the file actually have the shape we expect, this could be done with zod

Parsing the file with zod sounds nice, although we would have to think how to combine it with the `ConfigService` so the file wouldn't be validated twice.

>But instead of setting it on get, we'd need to update createFileStorage to accept a generic type State which could be then used in other methods.

👍

File: web/packages/teleterm/src/services/fileStorage/fileStorage.ts (hunk 1 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/fileStorage/fileStorage.ts]: I see your point, but are these flags really mutually exclusive?

>debounceWrites as a flag just doesn't make any sense when discardUpdatesWhenLoadingFileFailed is set to true because then debounceWrites has no effect

That's not entirely true. `debounceWrites` has no effect when `discardUpdatesWhenLoadingFileFailed` is set **and** an error occurred while loading the file. It will correctly debounce writes if the file has been loaded with no errors!

File: web/packages/teleterm/src/services/fileStorage/fileStorage.ts (hunk 6 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/fileStorage/fileStorage.ts]: True :(

File: web/packages/teleterm/src/main.ts [hunk 10]
```diff
@@ -97,16 +90,17 @@ function initializeApp(): void {
 
   app.on('will-quit', async event => {
     event.preventDefault();
+    const disposeMainProcess = async () => {
```
Comment (gzdunek) [inline web/packages/teleterm/src/main.ts:93]: I would put it as a top-level function, but it would require passing two arguments to it :/ I guess this is better?

File: web/packages/teleterm/src/services/fileStorage/fileStorage.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (ravicious) [reply web/packages/teleterm/src/services/fileStorage/fileStorage.ts]: I added a comment to `get` which links to this discussion.

File: web/packages/teleterm/src/services/fileStorage/fileStorage.ts (hunk 1 above)
Comment (ravicious) [reply web/packages/teleterm/src/services/fileStorage/fileStorage.ts]: You're right that they're indeed not mutually exclusive. Still, it was added to handle one specific situation which we could deal with differently with a different API.

File: web/packages/teleterm/src/ui/initUi.ts (hunk 2 above)
Comment (ravicious) [reply web/packages/teleterm/src/ui/initUi.ts]: > I like your suggestion, but perhaps without `the`? We don't have it other headers ('Encountered errors in config file').

Sounds good, I'd still use `the` in the description though.

File: web/packages/teleterm/src/main.ts (hunk 10 above)
Comment (ravicious) [reply web/packages/teleterm/src/main.ts:93]: That's fine I guess, it's only called once. I suppose using a promise also would be a good enough solution here.

**Review by ravicious — APPROVED** [review]

File: web/packages/teleterm/src/main.ts [hunk 11]
```diff
@@ -97,16 +90,17 @@ function initializeApp(): void {
...
-    appStateFileStorage.writeSync();
     globalShortcut.unregisterAll();
-    try {
-      await mainProcess.dispose();
-    } catch (e) {
-      logger.error('Failed to gracefully dispose of main process', e);
-    } finally {
-      app.exit();
-    }
+    await Promise.all([appStateFileStorage.write(), disposeMainProcess()]); // none of them can throw
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply web/packages/teleterm/src/main.ts:102]: That's theroretically possible, perhaps any other calls after `write()` should not be debounced? idk

File: web/packages/teleterm/src/ui/initUi.ts [hunk 12]
```diff
@@ -60,7 +60,7 @@ function notifyAboutConfigErrors(
       case 'file-loading': {
         notificationsService.notifyError({
           title: 'Failed to load config file',
-          description: `Using default config instead.\n${configError.error.toString()}`,
+          description: `Using the default config instead.\n${configError.error.toString()}`,
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply web/packages/teleterm/src/ui/initUi.ts]: Oh, we don't have to call `toString()`, as you said, it's done for us.

**Review by avatus — APPROVED** [review]
